### PR TITLE
Di component registry

### PR DESCRIPTION
**Pull request by maschmann**
On top of upgrade PR
* Inject ComponentBuilder besides Graph to decouple component instantiation
* restructure directories to current standards
* return $this as default for network and graph to allow chaining
* reworked examples

goal was to provide the means to integrate the library with e.g. a dependency injection container/ framework.

### CODE REVIEW

**Comment by maschmann** [comment]
Damn. Haven't tested with 5.4 - Class::class does not work in 5.4 -.-
Since there's no more active support for php 5.4 (http://php.net/supported-versions.php) I'd suggest to remove it from travis tests. Although the code works.

**Comment by hacfi** [comment]
Great to see some work being done on this! Don’t have much time to test this at the moment but I hope to give this a try at some point!

**Comment by maschmann** [comment]
@hacfi Hi, haven't read you for some time now ;-) just visit the channel again, will you?
Thanks for the encouragement - hopefully @bergie  has some time to merge the PR some time soon. before the PR becomes a complete rewrite :-P

**Comment by bergie** [comment]
@maschmann @hacfi to be honest, I haven't been doing any PHP in a while. So I'd be happy to give somebody else maintainer access to PhpFlo

cc/ @aretecode

File: README.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
-PhpFlo: Flow-based programming for PHP 5.3+
+PhpFlo: Flow-based programming for PHP 5
 ==============================================
 
-PhpFlo is a simple [flow-based programming](http://en.wikipedia.org/wiki/Flow-based_programming) implementation for PHP 5.3+. It is a PHP port of [NoFlo](https://github.com/bergie/noflo), a similar tool for Node.js. From WikiPedia:
+PhpFlo is a simple [flow-based programming](http://en.wikipedia.org/wiki/Flow-based_programming) implementation for PHP 5. It is a PHP port of [NoFlo](https://github.com/bergie/noflo), a similar tool for Node.js. From WikiPedia:
```
Comment (aretecode) [inline README.md]: If you changed from 5.3 to 5.5 it would be good if this could reflect that.

File: composer.json [hunk 2]
```diff
@@ -14,12 +13,18 @@
...
+        "php": ">=5.4.0",
+        "evenement/evenement": "~2.0"
+    },
+    "require-dev": {
+        "phpunit/phpunit": "<5.0"
     },
     "autoload": {
-        "psr-0": {
-            "PhpFlo": "src"
+        "psr-4": {
```
Comment (aretecode) [inline composer.json]: Much better!

File: README.md (hunk 1 above)
Comment (maschmann) [inline README.md]: Wouldn't it be better to use >=5.5 | 7.0 ? :-)

File: lib/PhpFlo/Component.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * @package PhpFlo
+ * @author Henri Bergius <henri.bergius@iki.fi>
+ */
+class Component implements ComponentInterface
+{
+    /**
+     * @var PortRegistry
+     */
+    private $_inPorts = null;
```
Comment (aretecode) [inline lib/PhpFlo/Component.php]: It would be cool if this adhered to http://www.php-fig.org/psr/psr-2/#4-2-properties

**Comment by aretecode** [comment]
@maschmann I really like this PR, I did a quick review, great modernization. It looks good to me @bergie. If you want to continue improving it, some next steps would be to put the components in another library pulled in by composer, and then https://github.com/noflo/noflo has some great improvements with brackets.

File: lib/PhpFlo/Component.php (hunk 3 above)
Comment (maschmann) [inline lib/PhpFlo/Component.php]: You're absolutely right! Damn. That's from my try to use magic methods to initialize the PortRegistry for in/out ports. But PSR-2... that's outright embarrassing to miss that O_o

**Comment by maschmann** [comment]
@aretecode @bergie moving the compoents to another library sounds like a good idea. before that I suggest merging both PRs since they kinda belong together.
I've built the initialization of the network component in a way so one could easily integrate with a DI container within a framework (since I mostly work with Symfony). I also created a [bundle](https://github.com/maschmann/PhpFloBundle) that integrate phpflo into symfony which only awaits this PR (currently using my fork for production, not cool)

**Comment by bergie** [comment]
@aretecode I've sent you an invite to this repo so you can handle the merging/releasing

**Comment by aretecode** [comment]
@maschmann in regards to https://github.com/bergie/phpflo/pull/16/commits/f276bee6daeea66a69e7062c7a0e4e0af586a94e
https://github.com/noflo/noflo/pull/469 might be considered.

**Pull request merged**